replace PyExecJS with PyMiniRacer?

sheppard opened the pull request:
Thanks for creating this project!  I am wondering if you would be interested in replacing the embedded PyExecJS environment detection with a dependency on [PyMiniRacer](https://github.com/sqreen/PyMiniRacer).

Advantages:
 * No implicit dependency on external binaries
 * In-memory variable bridging (no need for temporary JS files)
 * JS execution handled by existing lib, so less code to maintain

Disadvantages:
 * Requires ~6.6MB Python wheel as dependency
 * Many systems already have node installed, so this is somewhat redundant
 * PyMiniRacer is technically still in Pre-Alpha status, and doesn't work on Windows (yet)

Let me know what you think!

yetone commented:
Hi. First thank you for your contribution.

PyMiniRacer is an awesome project. But this is the first time I have heard about this project, so I do not know if it can be used in a production environment. 

Because this project in our company has a very important role, so whether it can be used for the production environment is my first consideration.

Performance and robustness are the points I want to consider. So I need your experience. 😀

sheppard commented:
Sounds good.  I will do a bit more testing and get back to you.

messense commented:
Why not support both? Let user decides which backend they want to use.

yetone commented:
@messense Looks like a good suggestion

sheppard commented:
Ok, I added back a node shim, removed the hard dependency on py-mini-racer, and implemented a few more tests.  The transformer will use PyMiniRacer if it is installed, otherwise it will fall back to node.  This can be manually specified with the `use_mini_racer` argument to the transform functions.  The name of this setting and/or the default precedence could certainly be changed.

```python
from babeljs import transformer

transformer.transform_string('const a = () => 233') # use default
transformer.transform_string('const a = () => 233', use_mini_racer=False) # use Node
transformer.transform_string('const a = () => 233', use_mini_racer=True) # use PyMiniRacer
```

To test performance and robustness, I customized the test suite to run on both backends and then compare the execution times.  Here are the summarized results from the [last run](https://travis-ci.org/yetone/babeljs-python/builds/287244675):

```
test_transform           : 13.29% faster (-73.0ms)
test_transform_string    : 9.86% faster (-55.0ms)
test_transform_vue_jsx   : 3.45% faster (-27.0ms)
test_invalid_js          : 29.65% faster (-178.0ms)
test_jquery              : 473.05% slower (+29990.0ms)
test_large_file          : 69.52% faster (-1158.0ms)
==== 12 passed in 49.63 seconds ====
```
In general, it looks like PyMiniRacer is a little faster for trivial test cases, but quite a bit slower when dealing with an actual library like jQuery. (Edit: this was due to returning the AST, see below)

sheppard commented:
Wait - if I disable returning the AST (which we aren't using anyway), it is much faster in both Node and PyMiniRacer.  Here's the updated summary:
```
test_transform           : 5.07% faster (-23.0ms)
test_transform_string    : 9.41% faster (-48.0ms)
test_transform_vue_jsx   : 6.01% faster (-38.0ms)
test_invalid_js          : 9.86% faster (-43.0ms)
test_jquery              : 0.15% slower (+5.0ms)
test_large_file          : 11.31% faster (-58.0ms)
==== 12 passed in 9.82 seconds ====
```

yetone commented on babeljs/jscontext.py:
```diff
@@ -0,0 +1,102 @@
+    pass
+
+
+def get_context(use_mini_racer=None):
```
`use_mini_racer=False` ?

sheppard replied:
The idea was to have `None` mean "autodetect".  Perhaps it would be clearer to make this a three-state variable instead of a boolean?

e.g.
```python
transformer.transform_string('const a = () => 233') # use default (autodetect)
transformer.transform_string('const a = () => 233', runtime='auto') # autodetect
transformer.transform_string('const a = () => 233', runtime='node') # use Node
transformer.transform_string('const a = () => 233', runtime='miniracer') # use PyMiniRacer
```

yetone replied:
@sheppard Since this, named `runtime` is better. `use_mini_racer` is more like a boolean value

yetone commented on babeljs/jscontext.py:
```diff
@@ -0,0 +1,102 @@
+        # Dump "eval" code to file, run node, and parse result. Makes some
+        # assumptions about code structure, which are safe in this case.
+        js_fd, js_filename = tempfile.mkstemp('.js')
+        result_fd, result_filename = tempfile.mkstemp('.json')
```
I think this will be a bit more intuitive and security:

```python
with tempfile.NamedTemporaryFile(suffix='.js') as js_fp:
    with tempfile.NamedTemporaryFile(suffix='.json') as res_fp:
        js_filename = js_fp.name
        res_filename = res_fp.name
        # balabala...

```

sheppard replied:
Sure, I will see if I can make it work this way.  I think my reason for doing it the other way was to make sure the subprocess has access to write to res_filename.  (It's more of an issue on Windows).

sheppard commented:
Ok, I made the suggested changes.  Let me know there is anything else or if you would like me to squash/rebase the commits.

yetone commented:
@sheppard need you squash :)

sheppard commented:
Ok, I squashed everything down to one commit.
